Unused Pattern Bindings (Unused Function Parameters) (#2022)
carbon-language/carbon-lang: proposals/p2022.md

Exceptions. See /LICENSE for license information.
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

[Pull request](https://github.com/carbon-language/carbon-lang/pull/2022

<!-- toc -->

## Table of contents

-   [Problem](#problem)
-   [Background](#background)
-   [Proposal](#proposal)
-   [Details](#details)
    -   [Examples](#examples)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)
    -   [Commented Names](#commented-names)
    -   [Only short form support with `_`](#only-short-form-support-with-_)
    -   [Anonymous, Named Identifiers](#anonymous-named-identifiers)

<!-- tocstop -->

## Problem

"How does a user of Carbon declare an unused pattern binding?"

Given that function parameters are a specific type of pattern binding, a more
specific question that will have the same answer is "How does a user of Carbon
declare an unused function parameter?"

Bindings that can be explicitly specified as unused allow authors of code to
clearly state that a value is not intended to be used and allow tools such as
compilers or linters to explicitly handle an argument that is specified as
unused, but later used.

> ## Background

Reviewer: Please include a link to https://github.com/carbon-language/carbon-lang/issues/476 for background, just as a back reference for any future research.

Author: Done. Also linked the section of the general design which describes what `_` means in the context of name binding.

addressed unused function parameters specifically with varying levels of clarity
for readers.

[Standard C++ Foundation's guidelines on unused parameters](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-unused)

[Google's C++ style guide recommendations on unused parameters](https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions)

[C++ `maybe_unused` attribute](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused)

[GCC C `unused` attribute](https://gcc.gnu.org/onlinedocs/gcc-4.7.2/gcc/Function-Attributes.html#index-g_t_0040code_007bunused_007d-attribute_002e-2640)

More generally, various languages use the underscore character (`_`) to signal
an unused binding or pattern.

[Rust `_` Patterns](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#ignoring-values-in-a-pattern)

[Golang Blank Identifiers](https://go.dev/ref/spec#Blank_identifier)

[Scala Wildcard Patterns](https://www.scala-lang.org/files/archive/spec/2.11/10-pattern-matching.html#variable-patterns)

[Crystal `_` Cases](https://crystal-lang.org/reference/1.5/syntax_and_semantics/case.html#underscore-allowed)

> ## Proposal

Reviewer: Can you comment on whether you think letting non-`unused` bindings go unused should be valid? It's fine for this to be an open question (e.g., https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0851.md#open-questions), I'd just like this proposal to be clear about the stance it's taking on that (or, whether it's not taking a stance, which would be fine too -- it's the kind of discussion that may not be resolved quickly).

I see arguments both ways -- it can be helpful when developing/debugging to not need to mark inputs as unused, but it can also lead to unused inputs becoming bugs.

Author: Included this in the latest changes. I very briefly and somewhat handwav-ily say there will be a diagnostic warning if a name binding isn't used. I'm not sure how much about diagnostics have already been specified or whether a statement like this needs to be backed up with much evidence about why this stance was taken. If it ends up requiring a fair bit more specification or justification I'm happy to just leave it as an open question.

Users can explicitly declare a value in a pattern as unused in two forms:

-   `_: Int` : Using the underscore token, `_`, in place of a name.

-   `unused size: Int` : Using the leading `unused` keyword followed by the
    name.

## Details

Introducing two syntaxes satisfies the desire to have a terse way to discard
values, but still provides authors with a more verbose, explicit syntax that
preserves the name.

Both approaches are unambiguous -- to human readers and authors as well as
programmatic interpretations. The inclusion of an explicit `unused` keyword
allows authors to preserve the name of a value for documentation purposes, while
still explicitly marking the value as discarded in an interpretable way to

> humans and programs alike.

Reviewer: One thing this proposal doesn't state is the actual behavior of `unused` bindings. It doesn't need to be much, but I'd suggest making this clear. e.g., one option would be: "Whereas `_` doesn't introduce a name, `unused` parameters are visible for name lookup but uses are invalid, including when they cause ambiguous name lookup errors."

A benefit of stating explicitly is that there's an alternative that `unused` could not add the name to name lookup, more consistent with `_`. This would be consistent with `/*foo*/` comments, but there's a possible advantage to "reserving" the name for later use (and avoiding accidents where there happens to be a matching name for lookup to find).

Author: I added a new section briefly describing `unused` in the context of name resolution. Pretty much used your quote here verbatim as it was explicit and clear. Let me know if more detail is needed.

# Unused Pattern Bindings (Unused Function Parameters)

<!--
Part of the Carbon Language project, under the Apache License v2.0 with LLVM
Exceptions. See /LICENSE for license information.
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

[Pull request](https://github.com/carbon-language/carbon-lang/pull/2022

<!-- toc -->

## Table of contents

-   [Problem](#problem)
-   [Background](#background)
-   [Proposal](#proposal)
-   [Details](#details)
    -   [The behavior of `unused` name bindings](#the-behavior-of-unused-name-bindings)
    -   [Examples](#examples)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)
    -   [Commented Names](#commented-names)
    -   [Only short form support with `_`](#only-short-form-support-with-_)
    -   [Anonymous, Named Identifiers](#anonymous-named-identifiers)

<!-- tocstop -->

> 

Reviewer 2: We'd like to include an abstract section as of  #1895 . It can match the summary in GitHub.

```suggestion
## Abstract

This proposal specifies how unused pattern bindings are written in the Carbon
programming language. This is a more general problem statement of "how do
users specify unused function parameters" as function parameter declarations
are a more specific form of pattern.

Related issue:
[#1996](https://github.com/carbon-language/carbon-lang/issues/1996).
```

Author: Done

## Problem

"How does a user of Carbon declare an unused pattern binding?"

Given that function parameters are a specific type of pattern binding, a more
specific question that will have the same answer is "How does a user of Carbon
declare an unused function parameter?"

Bindings that can be explicitly specified as unused allow authors of code to
clearly state that a value is not intended to be used and allow tools such as
compilers or linters to explicitly handle an argument that is specified as

> unused, but later used.

Reviewer 2: Long sentence should ideally be broken up.

Author: Updated

## Background

See
[the overall design](https://github.com/carbon-language/carbon-lang/blob/09e4417431b77c05ceb2a98dd38833276514a1ff/docs/design/README.md#binding-patterns)
for the current state of things with regards to name bindings and the underscore
character, `_`.

> See #476 for previous conversation on this topic.

Reviewer 2: ```suggestion
See
[issue #476: Optional argument names (unused arguments)](https://github.com/carbon-language/carbon-lang/issues/476)
for the previous conversation on this topic.
```

Author: Updated

Note: These are not exhaustive lists and not intended to be thoroughly
investigated, rather a brief overview of some prior art.

Various languages, style guides across languages, linters, and other tools have
addressed unused function parameters specifically with varying levels of clarity

> for readers.

Reviewer 2: The list of links below are all about C/C++.

```suggestion
C and C++ style guides, linters, and other tools have addressed unused
function parameters specifically with varying levels of clarity for readers.
```

Author: Updated

[Standard C++ Foundation's guidelines on unused parameters](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-unused)

[Google's C++ style guide recommendations on unused parameters](https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions)

[C++ `maybe_unused` attribute](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused)

> [GCC C `unused` attribute](https://gcc.gnu.org/onlinedocs/gcc-4.7.2/gcc/Function-Attributes.html#index-g_t_0040code_007bunused_007d-attribute_002e-2640)

Reviewer 2: Markdown list formatting:

```suggestion
-   [Standard C++ Foundation's guidelines on unused parameters](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-unused)
-   [Google's C++ style guide recommendations on unused parameters](https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions)
-   [C++ `maybe_unused` attribute](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused)
-   [GCC C `unused` attribute](https://gcc.gnu.org/onlinedocs/gcc-4.7.2/gcc/Function-Attributes.html#index-g_t_0040code_007bunused_007d-attribute_002e-2640)
```

Author: Updated

More generally, various languages use the underscore character (`_`) to signal
an unused binding or pattern.

[Rust `_` Patterns](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#ignoring-values-in-a-pattern)

[Golang Blank Identifiers](https://go.dev/ref/spec#Blank_identifier)

[Scala Wildcard Patterns](https://www.scala-lang.org/files/archive/spec/2.11/10-pattern-matching.html#variable-patterns)

> [Crystal `_` Cases](https://crystal-lang.org/reference/1.5/syntax_and_semantics/case.html#underscore-allowed)

Reviewer 2: In addition to making this a list, might as well throw Python/PyLint in there.

```suggestion
-   [Rust `_` Patterns](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#ignoring-values-in-a-pattern)
-   [Golang Blank Identifiers](https://go.dev/ref/spec#Blank_identifier)
-   [PyLint allows parameters starting with `_`, `ignored_`, or `unused_` to be unused](https://pylint.pycqa.org/en/latest/user_guide/configuration/all-options.html#ignored-argument-names)
-   [Scala Wildcard Patterns](https://www.scala-lang.org/files/archive/spec/2.11/10-pattern-matching.html#variable-patterns)
-   [Crystal `_` Cases](https://crystal-lang.org/reference/1.5/syntax_and_semantics/case.html#underscore-allowed)
```

Author: Updated

## Proposal

Users can explicitly declare a value in a pattern as unused in two forms:

> -   `_: Int` : Using the underscore token, `_`, in place of a name.

Reviewer 2: Please replace `Int` with `i32` in this entire document.

```suggestion
-   `_: i32` : Using the underscore token, `_`, in place of a name.
```

Author: Updated

Both approaches are unambiguous -- to human readers and authors as well as
programmatic interpretations. The inclusion of an explicit `unused` keyword
allows authors to preserve the name of a value for documentation purposes, while
still explicitly marking the value as discarded in an interpretable way to
humans and programs alike.

The `unused` keyword would be a new keyword in Carbon. This keyword would only
be valid when preceding a name in a pattern binding and the keyword would
tightly bind to the following name, disallowing specifying an entire sub-pattern
as `unused`.

### The behavior of `unused` name bindings

Names that are qualified with the `unused` keyword are visible for name lookup
but uses are invalid, including when they cause ambiguous name lookup errors. If
attempted to be used, a compiler error will be shown to the user, instructing
them to either remove the `unused` qualifier or remove the use.

The inverse, where a name is not qualified by the `unused` qualifier, but never
used, will cause the compiler to emit a warning diagnostic, informing the user
that a given name was not used and suggesting to either remove the binding or
mark it as unused.

> ### Examples

Reviewer 2: I find this section redundant. Can you combine the `_` and `unused` examples? I gave suggestions with the "function parameter" examples; please do the same for the others.

Author: Updated

> Discarded Name : Function Parameter

Reviewer 2: ```suggestion
Examples with an unused function parameter:
```

```carbon

> // Interface

Reviewer 2: The word "interface" has a specific, different meaning in Carbon

```suggestion
// Function declaration (may be in API file)
```

fn Sum(x: List(Int), size: Int) -> Int;

// Implementation that doesn't use the size parameter

> fn Sum(x: List(Int), _: Int) -> Int;

Reviewer 2: ```suggestion
fn Sum(x: List(Int), _: Int) -> Int { ... }
// or:
fn Sum(x: List(Int), unused size: Int) -> Int { ... }
```
This "or" is instead of the other "Function Parameter" example below.

```

Discarded Name : Case Binding

```carbon
fn Bar() -> (Int, Int);
fn Foo() -> Int {
  match (Bar()) {
    case (42, y: Int) => {
      y
    }

>     case (x: Int, _: Int => {

Reviewer 2: ```suggestion
    case (x: Int, _: Int) => {
```

      x
    }
  }
}
```

Discarded Name : Variable Binding

```carbon
fn Bar() -> (Int, Int);
fn Foo() -> Int {
  var (x: Int, _: Int) = Bar();
  return x;
}
```

> Unused Name : Function Parameter
> 
> ```carbon
> // Interface
> fn Sum(x: List(Int), size: Int) -> Int;
> 
> // Implementation that doesn't use the size parameter
> fn Sum(x: List(Int), unused size: Int) -> Int;
> ```
> 

Reviewer 2: ```suggestion
```

Unused Name : Case Binding

```carbon
fn Bar() -> (Int, Int);
fn Foo() -> Int {
  match (Bar()) {
    case (42, y: Int) => {
      return y;
    }
    case (x: Int, unused y: Int) => {
      return x;
    }
  }
}
```

Unused Name : Variable Binding

```carbon
fn Bar() -> (Int, Int);
fn Foo() -> Int {
  var (x: Int, unused y: Int) = Bar();
  return x;
}
```

## Rationale

> -   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)

Reviewer 2: ```suggestion
This proposal supports the Carbon goal of having
[code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write):

```

-   Carbon should not use symbols that are difficult to type, see, or
    differentiate from similar symbols in commonly used contexts.

> 

Reviewer 2: ```suggestion
```

-   Syntax should be easily parsed and scanned by any human in any development
    environment, not just a machine or a human aided by semantic hints from an
    IDE.
-   Explicitness must be balanced against conciseness, as verbosity and ceremony
    add cognitive overhead for the reader, while explicitness reduces the amount
    of outside context the reader must have or assume.

This proposal uses the underscore, `_`, character to denote an unused value, a
meaning used across various other programming languages. A lone underscore
character only has a single meaning in Carbon and will be unambiguous across
contexts.

Both syntaxes are easily read by humans, either by seeing the `_` chacter alone,
or by introducing a keyword that allows code to read naturally such as

> `unused size : Int`.

Reviewer 2: ```suggestion
`unused size: i32`.
```

The inclusion of two syntaxes allows authors to decide when they will favor

> conciseness or explicitness over the other.

Reviewer 2: ```suggestion
conciseness or explicitness.
```

> ## Alternatives considered

Reviewer 2: Another rule that was considered was saying any parameter named starting with an `_` is unused. See https://github.com/carbon-language/carbon-lang/issues/476#issuecomment-826068624 . Please include the alternative and the rationale why it was not chosen (both what was stated in the issue and in https://github.com/carbon-language/carbon-lang/pull/2022/files#r945934384 ).

Author: Added an alternative for this -- reading over #476 it seemed the main argument against is that its "clever". Let me know if I should add something more

Reviewer 2: Should also include something about https://github.com/carbon-language/carbon-lang/pull/2022/files#r945934384 : "reserving" the name for later use (and avoiding accidents where there happens to be a matching name for lookup to find).

Author: I didn't include anything along those lines because if names prefixed with `_` are discarded, there's nothing for them to ever conflict with or accidentally resolve to. Do you mean to include a pro that it can't lead to accidents?

Reviewer 2: It is a possible advantage of the `unused` keyword approach that you can reserve a name, and a potential disadvantage of the `_`-prefix approach that you can't.

Author: Gotcha -- thanks for the clarification. I'll update it this evening.

Reviewer 2: @chandlerc Maybe you could explain what was bad about being "clever" in https://github.com/carbon-language/carbon-lang/issues/476#issuecomment-826068624 ?

Reviewer 3: Good call, suggested more specific wording below.

-   [GCC C `unused` attribute](https://gcc.gnu.org/onlinedocs/gcc-4.7.2/gcc/Function-Attributes.html#index-g_t_0040code_007bunused_007d-attribute_002e-2640)

More generally, various languages use the underscore character (`_`) to signal
an unused binding or pattern.

-   [Rust `_` Patterns](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#ignoring-values-in-a-pattern)
-   [Golang Blank Identifiers](https://go.dev/ref/spec#Blank_identifier)
-   [PyLint allows parameters starting with `_`, `ignored_`, or `unused_` to be unused](https://pylint.pycqa.org/en/latest/user_guide/configuration/all-options.html#ignored-argument-names)
-   [Scala Wildcard Patterns](https://www.scala-lang.org/files/archive/spec/2.11/10-pattern-matching.html#variable-patterns)
-   [Crystal `_` Cases](https://crystal-lang.org/reference/1.5/syntax_and_semantics/case.html#underscore-allowed)

## Proposal

Users can explicitly declare a value in a pattern as unused in two forms:

-   `_: i32` : Using the underscore token, `_`, in place of a name.

-   `unused size: i32` : Using the leading `unused` keyword followed by the
    name.

## Details

> i32roducing two syntaxes satisfies the desire to have a terse way to discard

Reviewer 4: Looks like a search-and-replace for "int" -> "i32" got overenthusiastic here :-)

```suggestion
Introducing two syntaxes satisfies the desire to have a terse way to discard
```

Author: addressed in latest pushes

fn Sum(x: List(i32), unused size: i32) -> i32 { ... }
```

Examples with an unused variable

```carbon
fn Bar() -> (i32, i32);
fn Foo() -> i32 {
  var (x: i32, _: i32) = Bar();
  // or:
  var (x: i32, unused y: i32) = Bar();

  return x;
}
```

Examples with an unused case binding

```carbon
fn Bar() -> (i32, i32);
fn Foo() -> i32 {
  match (Bar()) {
    case (42, y: i32) => {
      return y;
    }
    case (x: i32, _: i32) => {
      return x;
    }
    // or:
    case (x: i32, unused y: i32) => {
      return x;
    }
  }
}
```

## Rationale

-   This proposal supports the Carbon goal of having
    [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
-   Carbon should not use symbols that are difficult to type, see, or
    differentiate from similar symbols in commonly used contexts.
-   Syntax should be easily parsed and scanned by any human in any development
    environment, not just a machine or a human aided by semantic hints from an
    IDE.
-   Explicitness must be balanced against conciseness, as verbosity and ceremony
    add cognitive overhead for the reader, while explicitness reduces the amount

>     of outside context the reader must have or assume.

Reviewer 4: These seem like they should be sub-points of the point above, rather than separate points at the same level.

Author: addressed in latest pushes

int Foo(int) { ... }
int Foo(int /*unused_name*/) { ... }
```

Advantages:

-   Consistency with C++

Disadvantages:

-   Carbon does not intend to support `/* */` comments, so this option is
    effectively a non-starter

### Only short form support with `_`

Carbon could provide only a single way to discard a value with the underscore,
`_`, token

```
// Function declaration (may be in API file)
fn Sum(x: List(i32), size: i32) -> i32;

// Implementation that doesn't use the size parameter
fn Sum(x: List(i32), _: i32) -> i32 { ... }
```

Advantages:

-   A smaller language, one less keyword, and a single way to write code

Disadvantages:

-   Less expressiveness, less documentation through names

### Named identifiers prefixed with `_`

Carbon could treat identifiers prefixed with `_` as unused identifiers and
discard their names.

```
// Function declaration (may be in API file)
fn Sum(x: List(i32), size: i32) -> i32;

// Implementation that doesn't use the size parameter
fn Sum(x: List(i32), _size: i32) -> i32 { ... }
```

Advantages:

-   Reuse of the underscore character in both short and long form bindings
-   Functionally similar to commented names in C++

Disadvantages:

> -   Interpreting the `_` character within an identifier name rather than using a
>     standalone token can be viewed as excessively clever

Reviewer 3: A suggested expansion on my "excessively clever" as it seems good to actually document something that covers all of this.

```suggestion
-   Tying the semantics if a name being unused to the particular spelling of the identifier (starting with a `_` character), even while it remains an identifier seems more subtle and indirect than necessary.
    -    We shouldn't use identifier spellings as a side-channel for semantic information.
    -    Semantics, including used vs. unused, should be conveyed in an orthogonal manner to the name rather than tying them together.
-   Would require the name to change when shifting between being used or unused, which could undermine some of its utility such as remaining for annotations, metaprogramming, or diagnostics.
```

Author: Thanks for the clarification. I updated to use this wording

    -   [Examples](#examples)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)
    -   [Commented names](#commented-names)
    -   [Only short form support with `_`](#only-short-form-support-with-_)
    -   [Named identifiers prefixed with `_`](#named-identifiers-prefixed-with-_)
    -   [Anonymous, named identifiers](#anonymous-named-identifiers)

<!-- tocstop -->

## Abstract

This proposal specifies how unused pattern bindings are written in the Carbon
programming language. This is a more general problem statement of "how do users
specify unused function parameters" as function parameter declarations are a
more specific form of pattern.

Related issue:
[#1996](https://github.com/carbon-language/carbon-lang/issues/1996).

## Problem

"How does a user of Carbon declare an unused pattern binding?"

Given that function parameters are a specific type of pattern binding, a more
specific question that will have the same answer is "How does a user of Carbon
declare an unused function parameter?"

Bindings that can be specified as unused makes code explicit and unambiguous.

> Authors of code can clearly state that a value is not intended to be used. Tools
> such as compilers or linters can explicitly handle an argument that is specified
> as unused, but later used.

Reviewer 5: ```suggestion
Authors of code can clearly state that a value is not intended to be used. Tools
such as compilers or linters can explicitly handle a parameter that is specified
as unused, but later used, or a parameter that is unused despite not being
specified as unused.
```

-   [PyLint allows parameters starting with `_`, `ignored_`, or `unused_` to be unused](https://pylint.pycqa.org/en/latest/user_guide/configuration/all-options.html#ignored-argument-names)
-   [Scala Wildcard Patterns](https://www.scala-lang.org/files/archive/spec/2.11/10-pattern-matching.html#variable-patterns)
-   [Crystal `_` Cases](https://crystal-lang.org/reference/1.5/syntax_and_semantics/case.html#underscore-allowed)

## Proposal

Users can explicitly declare a value in a pattern as unused in two forms:

-   `_: i32` : Using the underscore token, `_`, in place of a name.

-   `unused size: i32` : Using the leading `unused` keyword followed by the
    name.

## Details

Introducing two syntaxes satisfies the desire to have a terse way to discard
values, but still provides authors with a more verbose, explicit syntax that
preserves the name.

Both approaches are unambiguous -- to human readers and authors as well as
programmatic interpretations. The inclusion of an explicit `unused` keyword
allows authors to preserve the name of a value for documentation purposes, while
still explicitly marking the value as discarded in an interpretable way to
humans and programs alike.

> The `unused` keyword would be a new keyword in Carbon. This keyword would only

Reviewer 5: I would like to see a mention somewhere (perhaps "Alternatives considered") of using an attribute rather than a keyword -- whether it's saying "we've considered this and a keyword is the better choice" or to say "we don't have a design for attributes yet and we should consider whether we want to make this an attribute when we do" or even "once we have attributes this should definitely be one of those".

  match (Bar()) {
    case (42, y: i32) => {
      return y;
    }
    case (x: i32, _: i32) => {
      return x;
    }
    // or:
    case (x: i32, unused y: i32) => {
      return x;
    }
  }
}
```

## Rationale

-   This proposal supports the Carbon goal of having
    [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
    -   Carbon should not use symbols that are difficult to type, see, or
        differentiate from similar symbols in commonly used contexts.
    -   Syntax should be easily parsed and scanned by any human in any
        development environment, not just a machine or a human aided by semantic
        hints from an IDE.
    -   Explicitness must be balanced against conciseness, as verbosity and
        ceremony add cognitive overhead for the reader, while explicitness
        reduces the amount of outside context the reader must have or assume.

This proposal uses the underscore, `_`, character to denote an unused value, a
meaning used across various other programming languages. A lone underscore
character only has a single meaning in Carbon and will be unambiguous across
contexts.

> Both syntaxes are easily read by humans, either by seeing the `_` chacter alone,
> or by introducing a keyword that allows code to read naturally such as

Reviewer 5: ```suggestion
Both syntaxes are easily read by humans, either by seeing the `_` character
alone, or by introducing a keyword that allows code to read naturally such as
```

-   Would require the name to change when shifting between being used or unused,
    which could undermine some of its utility such as remaining for annotations,
    metaprogramming, or diagnostics.

### Anonymous, named identifiers

A potential syntax for naming unused bindings while retaining the underscore,
`_`, token is an optional name suffix following the underscore token.

```
// Function declaration (may be in API file)
fn Sum(x: List(i32), size: i32) -> i32;

// Implementation that doesn't use the size parameter
fn Sum(x: List(i32), _ size: i32) -> i32 { ... }
```

Advantages:

-   Reuse of the underscore token in both short and long form bindings

Disadvantages:

-   The underscore token, while consistent, may be less human readable than a
    dedicated keyword such as `unused`

### Attributes

While attributes aren't designed yet, there's a possibility that in the future,
the Carbon language will have some mechanism to attach metadata to parts of
source code to inform readers, compilers, and other tools. Its conceivable that
there could be an `unused` attribute which could be use to implement similar
semantics as the proposed `unused` keyword.

Advantages:

> -   Opens the inspection and subsequent actions taken to whatever ecosystem
>     and programmatic support is introduced by attributes.

Reviewer 5: ```suggestion
-   Opens the inspection and subsequent actions taken to whatever ecosystem and
    programmatic support is introduced by attributes.
```

## Details

> Intoducing two syntaxes satisfies the desire to have a terse way to discard

Reviewer 4: ```suggestion
Introducing two syntaxes satisfies the desire to have a terse way to discard
```

-   The underscore token, while consistent, may be less human readable than a
    dedicated keyword such as `unused`

### Attributes

While attributes aren't designed yet, there's a possibility that in the future,
the Carbon language will have some mechanism to attach metadata to parts of
source code to inform readers, compilers, and other tools. Its conceivable that
there could be an `unused` attribute which could be use to implement similar
semantics as the proposed `unused` keyword.

Advantages:

> -   Open the inspection and subsequent actions taken to the whatever ecosystem
>     and programmatic support introduced by attributes.

Reviewer 5: ```suggestion
-   Opens the inspection and subsequent actions taken to whatever ecosystem
    and programmatic support is introduced by attributes.
```
